compiler/rustc_middle/src/query/mod.rs

```diff
@@ -819,6 +819,32 @@ rustc_queries! {
         desc { |tcx| "comparing impl items against trait for `{}`", tcx.def_path_str(impl_id) }
     }
 
+    /// Maps from associated items on the impl specified by `impl_id`
+    /// to the corresponding associated item on the trait.
+    ///
+    /// For example, with the following code
+    ///
+    /// ```
+    /// struct Type {}
+    ///                         // DefId
+    /// trait Trait {           // trait_id
+    ///     fn f();             // trait_f
+    ///     fn g() {}           // trait_g
+    /// }
+    ///
+    /// impl Trait for Type {   // impl_id
+    ///     fn f() {}           // impl_f
+    ///     fn g() {}           // impl_g
+    /// }
+    /// ```
+    ///
+    /// The map returned for `tcx.impl_item_trait_item_ids(impl_id)` would be
+    ///`{ impl_f: trait_f, impl_g: trait_g }`
+    query impl_item_trait_item_ids(impl_id: DefId) -> &'tcx DefIdMap<DefId> {
```

Comment:
Why is this a query, and not just a helper function in dead.rs?

Reply:
I suggest using a local cache in dead.rs (in `MarkSymbolVisitor`?), and only after a perf run proving that it is necessary (preferably in a separate PR).